server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsContext.java

```diff
@@ -102,22 +105,43 @@ public void setChildInnerHits(Map<String, InnerHitSubContext> childInnerHits) {
             this.childInnerHits = new InnerHitsContext(childInnerHits);
         }
 
-        protected Weight createInnerHitQueryWeight() throws IOException {
-            final boolean needsScores = size() != 0 && (sort() == null || sort().sort.needsScores());
-            return context.searcher().createWeight(context.searcher().rewrite(query()),
+        protected Weight getInnerHitQueryWeight() throws IOException {
```

Comment:
This felt a bit hacky. We take a similar strategy in `FetchSourceContext#getFilter`.

Reply:
I think it's still hacky but slightly better.
There should be a way to share a state in `FetchSubPhase#hitExecute` though. 
If we plan to use `hitExecute` more, maybe we should provide a way to  share a state between hits ?
A factory could help for instance:
`````
Function<SearchContext, FetchSubPhase> fetchSubPhase;
`````

Reply:
This is a good idea, I do think we'll need a mechanism like this as we convert more phases to use `hitExecute`.